File: packages/rust/examples/sql_store.rs
#![allow(dead_code)]

//! Production-style SQL store for NEBULA — driver-agnostic template.
//!
//! The store is generic over a tiny [`SqlExecutor`] trait: implement it with the
//! driver of your choice (`postgres`, `rusqlite`, blocking `sqlx`) and you have
//! a spec-compliant store. That keeps this example dependency-free while
//! showing the exact SQL and semantics (schema: `docs/STORE.md`).
//!
//! What this file is really demonstrating:
//!
//! * [`RefreshTokenStore`] takes `&self`, so the store owns its synchronisation.
//!   Most Rust drivers want `&mut` for a query, so the connection lives behind a
//!   `Mutex` here; a pooled driver would check a connection out per call
//!   instead and drop the mutex entirely.
//! * `mark_rotated` and `revoke_if_active` are genuine compare-and-sets
//!   ([N-17], [N-18]): the status is part of the `WHERE` clause and the
//!   affected-row count is the return value. An `UPDATE` without it re-opens
//!   the race that forks a token family.
//! * Every failure is reported on the error channel ([N-20]) — nothing is
//!   swallowed into a protocol outcome, and an unrecognised `status` column
//!   fails closed rather than being read as `active`.
//! * Parameterised queries only; lookups keyed on the non-secret selector
//!   ([N-45]); rotated and revoked rows kept until the family's absolute
//!   deadline ([N-15]) — they are what makes replay detectable.

use nebula_token::{RefreshTokenStore, TokenRecord, TokenStatus};
use std::fmt;
use std::sync::Mutex;

/// A parameter value for a SQL statement.
pub enum SqlParam {
    Text(String),
    Int(i64),
    Null,
}

/// One row of `refresh_tokens`, in the canonical column order used below.
pub struct SqlRow {
    pub selector: String,
    pub verifier_hash: String,
    pub kid: String,
    pub family_id: String,
    pub generation: u32,
    pub user_id: String,
    pub device_id_hash: Option<String>,
    pub created_at: i64,
    pub family_expires_at: i64,
    pub idle_expires_at: i64,
    pub status: String,
    pub rotated_at: Option<i64>,
    pub replaced_by_selector: Option<String>,
}

/// Minimal driver abstraction. `execute` returns the affected-row count, which
/// is what makes the compare-and-sets below observable.
pub trait SqlExecutor {
    type Error;

    fn execute(&mut self, sql: &str, params: &[SqlParam]) -> Result<u64, Self::Error>;
    fn query_row(&mut self, sql: &str, params: &[SqlParam]) -> Result<Option<SqlRow>, Self::Error>;
}

/// Everything that can go wrong on the infrastructure channel ([N-20]).
#[derive(Debug)]
#[non_exhaustive]
pub enum SqlStoreError<E> {
    Driver(E),
    /// A previous caller panicked while holding the connection. Refusing is the
    /// fail-closed answer: we cannot know what the half-finished statement did.
    Poisoned,
    /// The `status` column holds something this version does not understand.
    /// Guessing `active` here would resurrect a revoked session.
    UnknownStatus(String),
}

impl<E: fmt::Display> fmt::Display for SqlStoreError<E> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SqlStoreError::Driver(e) => write!(f, "refresh-token store: {e}"),
            SqlStoreError::Poisoned => f.write_str("refresh-token store: connection poisoned"),
            SqlStoreError::UnknownStatus(s) => {
                write!(f, "refresh-token store: unknown status {s:?}")
            }
        }
    }
}

impl<E: fmt::Debug + fmt::Display> std::error::Error for SqlStoreError<E> {}

const COLS: &str = "selector, verifier_hash, kid, family_id, generation, user_id, device_id_hash, \
                    created_at, family_expires_at, idle_expires_at, status, rotated_at, replaced_by_selector";

pub struct SqlStore<E: SqlExecutor> {
    db: Mutex<E>,
}

impl<E: SqlExecutor> SqlStore<E> {
    pub fn new(db: E) -> Self {
        SqlStore { db: Mutex::new(db) }
    }

    fn with_db<T>(
        &self,
        f: impl FnOnce(&mut E) -> Result<T, E::Error>,
    ) -> Result<T, SqlStoreError<E::Error>> {
        let mut db = self.db.lock().map_err(|_| SqlStoreError::Poisoned)?;
        f(&mut db).map_err(SqlStoreError::Driver)
    }

    /// Operational helper: GC families past their absolute deadline. Never run
    /// anything narrower than this — deleting `rotated` rows early disables
    /// reuse detection ([N-15]).
    pub fn delete_expired(&self, now: i64) -> Result<u64, SqlStoreError<E::Error>> {
        self.with_db(|db| {
            db.execute(
                "DELETE FROM refresh_tokens WHERE family_expires_at <= $1",
                &[SqlParam::Int(now)],
            )
        })
    }
}

impl<E: SqlExecutor> RefreshTokenStore for SqlStore<E> {
    type Error = SqlStoreError<E::Error>;

    fn find_by_selector(&self, selector: &str) -> Result<Option<TokenRecord>, Self::Error> {
        let sql = format!("SELECT {COLS} FROM refresh_tokens WHERE selector = $1");
        let row = self.with_db(|db| db.query_row(&sql, &[SqlParam::Text(selector.to_string())]))?;
        row.map(row_to_record).transpose()
    }

    fn insert(&self, r: TokenRecord) -> Result<(), Self::Error> {
        let sql = format!(
            "INSERT INTO refresh_tokens ({COLS}) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13)"
        );
        self.with_db(|db| {
            db.execute(
                &sql,
                &[
                    SqlParam::Text(r.selector),
                    SqlParam::Text(r.verifier_hash),
                    SqlParam::Text(r.kid),
                    SqlParam::Text(r.family_id),
                    SqlParam::Int(i64::from(r.generation)),
                    SqlParam::Text(r.user_id),
                    r.device_id_hash.map_or(SqlParam::Null, SqlParam::Text),
                    SqlParam::Int(r.created_at),
                    SqlParam::Int(r.family_expires_at),
                    SqlParam::Int(r.idle_expires_at),
                    SqlParam::Text(r.status.as_str().to_string()),
                    r.rotated_at.map_or(SqlParam::Null, SqlParam::Int),
                    r.replaced_by_selector
                        .map_or(SqlParam::Null, SqlParam::Text),
                ],
            )
        })?;
        Ok(())
    }

    /// Compare-and-set ([N-17]): `status = $2` in the `WHERE` clause is the
    /// whole point, and `rows == 1` is the answer the engine acts on.
    fn mark_rotated(
        &self,
        selector: &str,
        from: TokenStatus,
        rotated_at: i64,
        replaced_by: &str,
    ) -> Result<bool, Self::Error> {
        let rows = self.with_db(|db| {
            db.execute(
                "UPDATE refresh_tokens SET status='rotated', rotated_at=$3, replaced_by_selector=$4 \
                 WHERE selector=$1 AND status=$2",
                &[
                    SqlParam::Text(selector.to_string()),
                    SqlParam::Text(from.as_str().to_string()),
                    SqlParam::Int(rotated_at),
                    SqlParam::Text(replaced_by.to_string()),
                ],
            )
        })?;
        Ok(rows == 1)
    }

    /// Compare-and-set ([N-18]).
    fn revoke_if_active(&self, selector: &str) -> Result<bool, Self::Error> {
        let rows = self.with_db(|db| {
            db.execute(
                "UPDATE refresh_tokens SET status='revoked' WHERE selector=$1 AND status='active'",
                &[SqlParam::Text(selector.to_string())],
            )
        })?;
        Ok(rows == 1)
    }

    /// `status <> 'revoked'` keeps the count honest, which is what makes the
    /// operation idempotent ([N-19]).
    fn revoke_family(&self, family_id: &str) -> Result<u64, Self::Error> {
        self.with_db(|db| {
            db.execute(
                "UPDATE refresh_tokens SET status='revoked' WHERE family_id=$1 AND status<>'revoked'",
                &[SqlParam::Text(family_id.to_string())],
            )
        })
    }

    fn revoke_user(&self, user_id: &str) -> Result<u64, Self::Error> {
        self.with_db(|db| {
            db.execute(
                "UPDATE refresh_tokens SET status='revoked' WHERE user_id=$1 AND status<>'revoked'",
                &[SqlParam::Text(user_id.to_string())],
            )
        })
    }
}

fn row_to_record<E>(r: SqlRow) -> Result<TokenRecord, SqlStoreError<E>> {
    let status = match r.status.as_str() {
        "active" => TokenStatus::Active,
        "rotated" => TokenStatus::Rotated,
        "revoked" => TokenStatus::Revoked,
        other => return Err(SqlStoreError::UnknownStatus(other.to_string())),
    };
    Ok(TokenRecord {
        selector: r.selector,
        verifier_hash: r.verifier_hash,
        kid: r.kid,
        family_id: r.family_id,
        generation: r.generation,
        user_id: r.user_id,
        device_id_hash: r.device_id_hash,
        created_at: r.created_at,
        family_expires_at: r.family_expires_at,
        idle_expires_at: r.idle_expires_at,
        status,
        rotated_at: r.rotated_at,
        replaced_by_selector: r.replaced_by_selector,
    })
}

fn main() {
    println!("This example is a template: implement SqlExecutor with your DB driver,");
    println!("then `NebulaEngine::new(Config::new(peppers, \"k1\", SqlStore::new(conn)))`.");
    println!("Wrap one refresh request in one transaction so the successor INSERT and");
    println!("the predecessor's compare-and-set UPDATE commit together ([N-22]).");
    println!("Schema and operational guidance: docs/STORE.md");
}

File: packages/go/examples/sqlstore/sqlstore.go
// Package sqlstore is a production-style SQL store for NEBULA built on the
// standard library's database/sql — driver-agnostic (works with pgx/stdlib,
// lib/pq, go-sqlite3, and MySQL with placeholder tweaks).
//
// What it demonstrates: parameterised queries only; lookups keyed on the
// non-secret selector ([N-45]); the two compare-and-set writes expressed as
// conditional UPDATEs whose RowsAffected is the return value ([N-17], [N-18]);
// rotated and revoked rows kept until the family's absolute deadline, because
// deleting them silently disables reuse detection ([N-15]); DeleteExpired for
// GC after that deadline.
//
// Wrap one refresh request in a single *sql.Tx at the call site so that the
// insert of the successor and the mark-rotated of the predecessor commit
// atomically ([N-22]); the engine's compensation in [N-34] step 5 covers the
// deployments that cannot.
//
// Schema: docs/STORE.md. This package lives under examples/ and is not part of
// the published module surface.
package sqlstore

import (
	"context"
	"database/sql"
	"errors"

	nebulatoken "github.com/nebula-token/nebula-token/packages/go"
)

const cols = `selector, verifier_hash, kid, family_id, generation, user_id, device_id_hash,
	created_at, family_expires_at, idle_expires_at, status, rotated_at, replaced_by_selector`

// Queryer is satisfied by *sql.DB and *sql.Tx alike.
type Queryer interface {
	QueryRowContext(ctx context.Context, query string, args ...any) *sql.Row
	ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error)
}

// Store implements nebulatoken.RefreshTokenStore over database/sql.
type Store struct{ DB Queryer }

func New(db Queryer) *Store { return &Store{DB: db} }

// Compile-time proof that the six methods still match the contract.
var _ nebulatoken.RefreshTokenStore = (*Store)(nil)

func (s *Store) FindBySelector(ctx context.Context, selector string) (*nebulatoken.TokenRecord, error) {
	var r nebulatoken.TokenRecord
	var deviceIDHash, replacedBy sql.NullString
	var rotatedAt sql.NullInt64
	var status string

	err := s.DB.QueryRowContext(ctx,
		`SELECT `+cols+` FROM refresh_tokens WHERE selector = $1`, selector,
	).Scan(&r.Selector, &r.VerifierHash, &r.Kid, &r.FamilyID, &r.Generation, &r.UserID,
		&deviceIDHash, &r.CreatedAt, &r.FamilyExpiresAt, &r.IdleExpiresAt,
		&status, &rotatedAt, &replacedBy)
	if errors.Is(err, sql.ErrNoRows) {
		// A missing row is a protocol outcome (NOT_FOUND), not an
		// infrastructure failure ([N-20]).
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	// NULL stays NULL: an unbound family and a family bound to the empty device
	// identifier are different things ([N-25]), and a record rotated at Unix
	// time 0 is not an unrotated one.
	if deviceIDHash.Valid {
		r.DeviceIDHash = &deviceIDHash.String
	}
	if replacedBy.Valid {
		r.ReplacedBySelector = &replacedBy.String
	}
	if rotatedAt.Valid {
		r.RotatedAt = &rotatedAt.Int64
	}
	r.Status = nebulatoken.TokenStatus(status)
	return &r, nil
}

func (s *Store) Insert(ctx context.Context, r *nebulatoken.TokenRecord) error {
	_, err := s.DB.ExecContext(ctx,
		`INSERT INTO refresh_tokens (`+cols+`)
		 VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13)`,
		r.Selector, r.VerifierHash, r.Kid, r.FamilyID, r.Generation, r.UserID,
		nullString(r.DeviceIDHash), r.CreatedAt, r.FamilyExpiresAt, r.IdleExpiresAt,
		string(r.Status), nullInt64(r.RotatedAt), nullString(r.ReplacedBySelector))
	return err
}

// MarkRotated is the compare-and-set of [N-17]: the status predicate is part of
// the WHERE clause and the affected-row count is the answer. An unconditional
// UPDATE that discarded sql.Result would report success to both of two
// concurrent refreshes and fork the family into two live lineages.
func (s *Store) MarkRotated(ctx context.Context, selector string, fromStatus nebulatoken.TokenStatus, rotatedAt int64, replacedBySelector string) (bool, error) {
	res, err := s.DB.ExecContext(ctx,
		`UPDATE refresh_tokens SET status='rotated', rotated_at=$2, replaced_by_selector=$3
		 WHERE selector=$1 AND status=$4`,
		selector, rotatedAt, replacedBySelector, string(fromStatus))
	if err != nil {
		return false, err
	}
	return changed(res)
}

// RevokeIfActive is the compare-and-set of [N-18].
func (s *Store) RevokeIfActive(ctx context.Context, selector string) (bool, error) {
	res, err := s.DB.ExecContext(ctx,
		`UPDATE refresh_tokens SET status='revoked' WHERE selector=$1 AND status='active'`, selector)
	if err != nil {
		return false, err
	}
	return changed(res)
}

// RevokeFamily returns the number of records it changed ([N-19]). The
// `status <> 'revoked'` predicate is what makes the count idempotent: revoking
// an already-revoked family reports 0, not its size.
func (s *Store) RevokeFamily(ctx context.Context, familyID string) (int, error) {
	return s.revokeWhere(ctx, `family_id=$1`, familyID)
}

// RevokeUser returns the number of records it changed ([N-19]).
func (s *Store) RevokeUser(ctx context.Context, userID string) (int, error) {
	return s.revokeWhere(ctx, `user_id=$1`, userID)
}

func (s *Store) revokeWhere(ctx context.Context, predicate string, arg any) (int, error) {
	res, err := s.DB.ExecContext(ctx,
		`UPDATE refresh_tokens SET status='revoked' WHERE `+predicate+` AND status <> 'revoked'`, arg)
	if err != nil {
		return 0, err
	}
	n, err := res.RowsAffected()
	if err != nil {
		// A driver that cannot count rows cannot support this contract:
		// reporting 0 would understate a revocation, reporting a guess would
		// overstate one. Fail closed instead ([N-20]).
		return 0, err
	}
	return int(n), nil
}

// DeleteExpired is an operational helper: collect families past their absolute
// deadline. Nothing may be deleted before it ([N-15]).
func (s *Store) DeleteExpired(ctx context.Context, now int64) (int64, error) {
	res, err := s.DB.ExecContext(ctx, `DELETE FROM refresh_tokens WHERE family_expires_at <= $1`, now)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

func changed(res sql.Result) (bool, error) {
	n, err := res.RowsAffected()
	if err != nil {
		return false, err
	}
	return n > 0, nil
}

func nullString(s *string) any {
	if s == nil {
		return nil
	}
	return *s
}

func nullInt64(n *int64) any {
	if n == nil {
		return nil
	}
	return *n
}
